src/moneypenny/handlers/external/agent.py

```diff
@@ -68,20 +72,21 @@ async def commission_agent(request: web.Request) -> web.Response:
         ),
     )
     moneypenny = request.config_dict["moneypenny"]
-    username = body["username"]
+    username = body["token"]["data"]["uid"]
     await moneypenny.execute_order(verb="post", dossier=body)
     return web.HTTPAccepted(text=f"Commissioning {username}")
 
 
-
 @routes.delete("/{username}")
 async def retire_agent(request: web.Request) -> web.Response:
     """DELETE /moneypenny/{username}
 
     Perform deprovisioning steps for the given username.  Returns a 202 once
     the request has been accepted.
     """
```

Comment:
If you're going to have a DELETE, the GET on `/{username}` needs to return success for all provisioned users, not just the ones with pods running, or the semantics get screwy (you can successfully DELETE a resource that returns 404).

I don't think DELETE semantics are what you want, since Moneypenny has no idea what users exist. Unless you're going to add knowledge to Moneypenny of what users exist (which doesn't seem right), I think deprovisioning should be a POST to some other route. It's a command, not a resource management instruction, so REST says that should be a POST. That may argue for not putting the POST for provisioning at the top level either; my inclination would be to have `/provision` and `/deprovision` routes and not try to be all RESTful here since REST semantics don't really fit (you're not managing a resource collection).

Reply:
Yeah, I think the symantics here are a bit confusing.  First off, passing down verb as far as as the kubernetes layer seems confusing.  It's a terrible word even for HTTP, and worse when we are using it.  We could call it operation, or task or anything like that.  Pick a pun, just not verb.

That being said, you might want to think of this more like a job system, like there's a /provision endpoint where you post the body, and it hands you a URL that is /provision/cbanek-123456 randomnumber or uuid or something.  That way you can check the status after it's finished and hold onto it until the caller calls DELETE on /provision/cbanek-123456 to cleanup.  The resources are actually what's happening, since what if a job gets stuck or takes a long time, etc.  Also, in the future when there may be more than one of these jobs running per user or those symantics come into play it would probably make sense to treat this more as a job system.

At the http layer have each URL be a job like /provision /deprovision /adminjob2 /takeoutback etc.  Then each of those can have things that are going.  You could just hold it in memory.  Then at the moneypenny layer it figures out the yaml or the resources, and passes that into the kubernetes layers?

Reply:
Changed to multiple endpoints, one per action, with a POST.

We actually only want one operation going at a time for any given user.  If a user receives another request while it has one being processed, you get back a 409 Conflict.